bootloader/bootloader.c

```diff
@@ -139,11 +139,15 @@ void __attribute__((noreturn)) __stack_chk_fail(void)
 
 int main(void)
 {
-	__stack_chk_guard = random32();
 	setup();
 	memory_protect();
 	oledInit();
 
+	// to be extra careful and heed the STM32F205xx Reference manual, Section 20.3.1
+	// we don't use the first random number generated after setting the RNGEN bit in setup
+	random32();
+	__stack_chk_guard = random32(); // this supports compiler provided unpredictable stack protection checks
```

Comment:
the call to random32 that discards the result does not have to come after memory_protect and oledInit. calling it later only avoids any possibly (although unlikely) perceptible delay before init'ing the screen. this is what was discussed two comments above (i had both of these lines just after the call to setup). one could also argue that maybe we should not progress to init the screen until the rng is generating quickly/successfully. it may be useful to have the discarding random32 call in setup() and setupApp(). doing that makes sense to me. @prusnak is there a reason why the demo.c code does not, or could not, call setupApp near the beginning of its main function? if it was there this refactor would be complete.

changing the stack guard from within a function that uses it causes that function to fail on return though because the hidden value pushed on the stack (by the compiler generated code) won't match anymore. so the second line can't be factored into setup and setupApp.